### ISSUE TITLE
Cannot plot more than 3000 data point.

### ISSUE THREAD

**Issue by Doopin** [entronad/graphic#247]
Thanks for this nice graphing tool.
Please any idea why I cannot plot more than 3.000 data point even though I setup to 10000 per group? I have the same issue if I use one LineMark only.
Below are part of my code. Can you please tell me whats wrong?

```
@override
void initState() {
  data = [
    {'index': 0, 'value': 0, 'group': 1},
    {'index': 0, 'value': 0, 'group': 2},
  ];

  for (int ii = 0; ii < 11000; ii++) {
    data.add({'index': ii, 'value': rdm.nextInt(1000), 'group': 1});
    data.add({
      'index': ii,
      'value': (rdm.nextInt(1000) - rdm.nextInt(100)),
      'group': 2
    });
    counter++;
  }

  super.initState();
}
```

```
Chart(
    rebuild: true,
    changeData: true,
    data: data,
    variables: {
      'index': Variable(
        accessor: (Map map) => map['index'] as num,
        scale: LinearScale(tickCount: 7),
      ),
      'value': Variable(
        accessor: (Map map) => map['value'] as num,
        scale: LinearScale(),
      ),
      'group': Variable(
        accessor: (Map map) => map['group'].toString(),
      ),
    },
    marks: [
      LineMark(
          position:
              Varset('index') * Varset('value') / Varset('group'),
          shape: ShapeEncode(value: BasicLineShape(smooth: true)),
          size: SizeEncode(value: 0.5),
          color: ColorEncode(
            variable: 'group',
            values: [
              const Color(0xff5470c6),
              const Color(0xff91cc75),
            ],
          )),
    ],
    axes: [
      Defaults.horizontalAxis,
      Defaults.verticalAxis,
    ],
    coord: RectCoord(
      horizontalRangeUpdater: Defaults.horizontalRangeEvent,
    ),
    selections: {
      'touchMove': PointSelection(
        on: {
          GestureType.scaleUpdate,
          GestureType.tapDown,
          GestureType.longPressMoveUpdate
        },
        dim: Dim.x,
      )
    },
    tooltip: TooltipGuide(
      followPointer: [false, true],
      align: Alignment.topLeft,
      offset: const Offset(-10, -10),
    ),
    crosshair: CrosshairGuide(followPointer: [false, true]),
  )
```

**Comment by entronad** [comment]
Of course, I'd be happy to help you with that. Thank you for using Graphic and for providing such a detailed report with code and a screenshot—it's incredibly helpful for diagnosing issues.

This is an AI-generated solution. Please let me know if it resolves your problem! Your feedback is valuable for improving the tool and its support.

---

### Why You're Seeing Fewer Than 10,000 Points

The behavior you're observing is not a bug, but rather a performance optimization feature directly related to the **line smoothing** you've enabled.

The core of the issue lies here:

```dart
shape: ShapeEncode(value: BasicLineShape(smooth: true)),
```

When you set `smooth: true`, Graphic uses a spline algorithm to draw a visually pleasing, curved line instead of a series of straight segments. Calculating splines is computationally more intensive than drawing simple straight lines. To ensure your application remains responsive and avoids performance bottlenecks or freezing with very large datasets, **Graphic automatically applies a data sampling algorithm when line smoothing is active**.

This sampling reduces the number of points used to render the smoothed line, which is why you see a cutoff around 3,000 data points instead of the full 11,000 per group that you've generated. The goal is to provide a visually representative curve without the performance cost of processing every single point.

### Suggestions and Solutions

You have a couple of excellent options to plot all your data points, depending on your visualization goals.

#### 1. (Recommended) Disable Line Smoothing for High-Density Data

For a dataset with over 10,000 points, the visual difference between a smoothed line and a regular line is often negligible. The sheer density of the points typically creates a smooth appearance on its own.

By disabling smoothing, you bypass the sampling algorithm, which will allow all your data points to be rendered and significantly improve performance.

**💡 How to implement:**

Simply set `smooth` to `false` or remove the `shape` property from your `LineMark` entirely, as a non-smoothed line is the default.

**Code Example:**

```dart
Chart(
  // ... all your other properties
  marks: [
    LineMark(
      position: Varset('index') * Varset('value') / Varset('group'),
      size: SizeEncode(value: 0.5),
      color: ColorEncode(
        variable: 'group',
        values: [
          const Color(0xff5470c6),
          const Color(0xff91cc75),
        ],
      ),
      // Option 1: Explicitly set smooth to false
      shape: ShapeEncode(value: BasicLineShape(smooth: false)), 
      // Option 2 (Simpler): Remove the shape property entirely.
    ),
  ],
  // ... rest of your chart properties
)
```

This is the most direct solution and should resolve the issue immediately.

---

#### 2. Manually Aggregate Your Data Before Plotting

If a smoothed line showing the general trend is essential, and you don't need to see every single fluctuation, a better approach for very large datasets is to perform data aggregation yourself.

Instead of feeding 11,000 raw points into the chart, you can pre-process the data to create a smaller, more manageable dataset that represents the overall trend. For example, you could group the data by an index range and calculate the average `value` for each range.

This approach gives you full control over the level of detail and guarantees excellent performance, even with smoothing enabled on the smaller, aggregated dataset.

**Conceptual Example of Aggregation:**

You could write a function to transform your raw `data` list:

```dart
List<Map<String, dynamic>> aggregateData(List<Map<String, dynamic>> rawData, int windowSize) {
  // Logic to group data by 'group'
  // Then, for each group, iterate through the data in chunks of 'windowSize'
  // Calculate the average 'index' and 'value' for each chunk
  // Return the new, smaller list of aggregated data points
  // ...
  return aggregatedData;
}

// In your widget:
final aggregated = aggregateData(data, 100); // Aggregate 11,000 points into 110

Chart(
  data: aggregated, // Use the smaller dataset
  // ...
  marks: [
    LineMark(
      // You can now use smooth: true with good performance
      shape: ShapeEncode(value: BasicLineShape(smooth: true)),
      // ...
    )
  ]
  // ...
)
```

This is a more advanced technique but is standard practice in data visualization for handling massive datasets effectively.

I hope this explanation is clear and helps you create the chart you envision! For more details on the available properties, you can always refer to the official documentation:

* **Graphic Documentation:** [https://pub.dev/documentation/graphic/latest/graphic/](https://pub.dev/documentation/graphic/latest/graphic/)

Let me know how it goes!
